### DISCUSSION TITLE
Specify related objects by attribute during CSV import. Case #3147

### DISCUSSION THREAD

**Discussion by kevintedder** [netbox-community/netbox#11988] [Q&A]
Issue #3147 answers my needs as I have duplicate named devices. However when I follow the details in this case it does not work.

The CSV file is follows:
address,status,virtual_machine.site,virtual_machine.name,interface
192.168.1.1/30,active,SITE_A,Server_1,Net1

Error message: 
Row 1: Expected 4 columns but found 5

If I add another field
address,status,virtual_machine.site,virtual_machine.name,interface,dns_name
192.168.1.1/30,active,SITE_A,Server_1,Net1

Error message: 
Row 1 virtual_machine: Object not found
Row 1 interface : Object not found

Using  v3.3.2

What am I doing wrong?

**Comment by mtinberg** [comment]
virtual_machine.site and virtual_machine.name are not column names, I believe this is being interpreted as virtual_machine column twice, which is why it is seen a 4 columns and not 5.  You probably want "site" and "name", but the import page lists the expected field names.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: kevintedder ***@***.***>
Sent: Wednesday, March 15, 2023 11:52 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Specify related objects by attribute during CSV import. Case #3147 (Discussion #11988)


Issue #3147<https://github.com/netbox-community/netbox/issues/3147> answers my needs as I have duplicate named devices. However when I follow the details in this case it does not work.

The CSV file is follows:
address,status,virtual_machine.site,virtual_machine.name,interface
192.168.1.1/30,active,SITE_A,Server_1,Net1

Error message:
Row 1: Expected 4 columns but found 5

If I add another field
address,status,virtual_machine.site,virtual_machine.name,interface,dns_name
192.168.1.1/30,active,SITE_A,Server_1,Net1

Error message:
Row 1 virtual_machine: Object not found
Row 1 interface : Object not found

Using v3.3.2

What am I doing wrong?

**Comment by candlerb** [comment]
Yes it's a weird error message, when clearly there are 5 CSV entries in the header and the row (confirmed here).

#3147 is very old, and I've never imported a CSV file in the way you show; I also don't see any test cases for the format you're using in `netbox/ipam/tests/test_views.py` - or in `netbox/dcim/tests/test_views.py` except for one instance of `vrf.pk`

**Reply by kevintedder** [reply]
#3147 was included and merged in Github. That would suggest that this feature is now available, unless it was only for DCIM/Devices, not for Virtual Machines (which would be disappointing).   I'll just have to enter the interfaces on 30+ VM's manually - :(

**Reply by candlerb** [reply]
What I was trying to say was:
* the functionality in [#3147](https://github.com/netbox-community/netbox/issues/3147) was not really defined in the ticket, nor in the [documentation](https://docs.netbox.dev/en/stable/getting-started/populating-data/#bulk-import-csvyaml).
* there are no test cases for using the way you *think* it should be used, which I think comes from @jeremystretch's proposal/idea in [this comment](https://github.com/netbox-community/netbox/issues/3147#issuecomment-565473082).
* there were other ideas for how it _might_ work, such as in the [initial post](https://github.com/netbox-community/netbox/issues/3147#issue-440199917) on that ticket.

So either it never worked in the way you think; or it did initially, but it broke because there are no tests.

The change has a [long PR](https://github.com/netbox-community/netbox/pull/4564) which completely reworks CSV import, and I can't see how (or if) it is intended to identify arbitrary related objects via multiple attributes of that object.

In [this commit](https://github.com/netbox-community/netbox/pull/4564/commits/61ae4be16a6ebd0e3741d491ebbbf6baf973eacb) I can see tests with single columns `vrf.name`, `vrf.pk`, `vrf.rd`, i.e. using *one* arbitrary attribute to identify a related object.

Furthermore, the [data structure returned](https://github.com/netbox-community/netbox/pull/4564/commits/61ae4be16a6ebd0e3741d491ebbbf6baf973eacb#diff-40028b1fe3ef399489926b6c477ceb9887c2d4bbac42e0e6b2702158ed5946a8R325-R327) implies that only a single identifying attribute can be used:

```
output = (
            {'address': None, 'status': None, 'vrf': 'name'},
                                              ^^^^^^^^^^^^^
```

i.e. it's not `'vrf': ['name']` or `'vrf': set('name')`.

EDIT: I suspect that's the source of the problem you see. I theorise that `{'virtual_machine': 'name'}` has overwritten `{'virtual_machine': 'site'}` so it thinks there's only one column heading instead of two.  If so, I think it should at least report this condition as an error, rather than silently dropping a column.  Making it work the way you want would be a significant rewrite.

I do see tests cases with columns "rackgroup" and "rack", or "site" and "device", where you're implicitly constraining the set of racks or devices to those within a given rackgroup or site.  But they don't say "device.site" and "device.name", for example.

**Reply by candlerb** [reply]
Raised as #11990.
